File: backend/services/tmdb.js
/**
 * tmdb.js — TMDB API client (v3 endpoints, Bearer token auth)
 *
 * Auth priority:
 *   1. TMDB_READ_ACCESS_TOKEN (Bearer JWT — preferred)
 *   2. TMDB_API_KEY            (legacy api_key query param — fallback)
 */
import dotenv from 'dotenv';
dotenv.config();

const BASE = 'https://api.themoviedb.org/3';

/* ------------------------------------------------------------------ */
/* Auth + fetch wrapper                                               */
/* ------------------------------------------------------------------ */

function buildRequest() {
  const bearer = process.env.TMDB_READ_ACCESS_TOKEN;
  const apiKey  = process.env.TMDB_API_KEY;

  if (bearer) {
    return {
      headers: { Authorization: `Bearer ${bearer}`, 'Content-Type': 'application/json' },
      extraParams: {},
    };
  }
  if (apiKey) {
    return { headers: {}, extraParams: { api_key: apiKey } };
  }
  throw new Error('No TMDB credentials: set TMDB_READ_ACCESS_TOKEN or TMDB_API_KEY in .env');
}

async function tmdbGet(pathname, queryParams = {}) {
  const { headers, extraParams } = buildRequest();
  const url = new URL(`${BASE}${pathname}`);
  for (const [k, v] of Object.entries({ ...extraParams, ...queryParams })) {
    if (v !== undefined && v !== null && v !== '') url.searchParams.set(k, String(v));
  }
  const res = await fetch(url.toString(), { headers });
  if (!res.ok) {
    const body = await res.text().catch(() => '');
    throw new Error(`HTTP ${res.status} — ${body.slice(0, 120)}`);
  }
  return res.json();
}

/* ------------------------------------------------------------------ */
/* Shared helpers                                                     */
/* ------------------------------------------------------------------ */

const BOLLYWOOD_LANGS   = new Set(['hi', 'te', 'ta', 'ml', 'kn']);
const EUROPEAN_COUNTRIES = new Set(['GB','FR','ES','IT','DE','SE','DK','NO','NL']);

function detectMarket(productionCountries = [], originalLanguage = '') {
  const countries = productionCountries.map((c) => c.iso_3166_1);
  const lang = originalLanguage || '';

  if (countries.includes('US') && lang === 'en') return 'hollywood';
  if (countries.includes('IN') && BOLLYWOOD_LANGS.has(lang)) return 'bollywood';
  if (countries.includes('KR')) return 'korean';
  if (countries.includes('JP')) return 'japanese';
  if (countries.some((c) => EUROPEAN_COUNTRIES.has(c))) return 'european';
  // US film with non-English language — route by language
  if (countries.includes('US')) {
    if (BOLLYWOOD_LANGS.has(lang)) return 'bollywood';
    if (lang === 'ko') return 'korean';
    if (lang === 'ja') return 'japanese';
    if (['fr','de','es','it'].includes(lang)) return 'european';
  }
  return 'hollywood';
}

function mapStatus(tmdbStatus = '') {
  switch (tmdbStatus.toLowerCase()) {
    case 'released':        return 'released';
    case 'in production':
    case 'post production': return 'in_production';
    case 'planned':
    case 'announced':       return 'upcoming';
    default:                return 'upcoming';
  }
}

function extractMpaaRating(releaseDates) {
  const results = releaseDates?.results ?? [];
  // US theatrical certification first
  const usEntry = results.find((r) => r.iso_3166_1 === 'US');
  if (usEntry) {
    const cert = (usEntry.release_dates ?? [])
      .filter((r) => r.type === 3 || r.type === 2)
      .find((r) => r.certification)?.certification;
    if (cert) return cert;
    // Any US cert as fallback
    const anyCert = (usEntry.release_dates ?? []).find((r) => r.certification)?.certification;
    if (anyCert) return anyCert;
  }
  // Any other country
  for (const entry of results) {
    const cert = (entry.release_dates ?? []).find((r) => r.certification)?.certification;
    if (cert) return cert;
  }
  return null;
}

function extractTrailerUrl(videos) {
  const results = videos?.results ?? [];
  const official = results.find(
    (v) => v.type === 'Trailer' && v.official === true && v.site === 'YouTube'
  );
  if (official) return `https://www.youtube.com/watch?v=${official.key}`;
  const anyYT = results.find((v) => v.type === 'Trailer' && v.site === 'YouTube');
  if (anyYT) return `https://www.youtube.com/watch?v=${anyYT.key}`;
  return null;
}

function today() {
  return new Date().toISOString().split('T')[0];
}
function dateOffset(months = 0, years = 0) {
  const d = new Date();
  d.setMonth(d.getMonth() + months);
  d.setFullYear(d.getFullYear() + years);
  return d.toISOString().split('T')[0];
}

// Language + country params per market (for discover endpoints).
function marketDiscoverParams(market) {
  switch (market) {
    case 'bollywood': return { with_original_language: 'hi|te|ta|ml|kn' };
    case 'korean':    return { with_original_language: 'ko' };
    case 'japanese':  return { with_original_language: 'ja' };
    case 'european':  return { with_original_language: 'fr|de|es|it' };
    case 'hollywood':
    default:          return { with_original_language: 'en' };
  }
}

/* ------------------------------------------------------------------ */
/* 1. searchFilm(query)                                               */
/* ------------------------------------------------------------------ */

export async function searchFilm(query) {
  try {
    const data = await tmdbGet('/search/movie', {
      query,
      include_adult: false,
    });
    return (data.results ?? []).slice(0, 5).map((r) => ({
      tmdb_id:      r.id,
      title:        r.title,
      release_date: r.release_date ?? null,
      poster_path:  r.poster_path ?? null,
      overview:     r.overview ?? null,
    }));
  } catch (err) {
    console.error('[TMDB ERROR] searchFilm:', err.message);
    return null;
  }
}

/* ------------------------------------------------------------------ */
/* 2. getFilmDetails(tmdb_id)                                         */
/* ------------------------------------------------------------------ */

export async function getFilmDetails(tmdbId) {
  try {
    const d = await tmdbGet(`/movie/${tmdbId}`, {
      append_to_response: 'credits,release_dates,videos',
    });

    const credits      = d.credits ?? {};
    const releaseDates = d.release_dates ?? {};
    const videos       = d.videos ?? {};

    const director = (credits.crew ?? []).find((c) => c.job === 'Director') ?? null;
    const cast_top5 = (credits.cast ?? []).slice(0, 5).map((c) => ({
      name:           c.name,
      tmdb_person_id: c.id,
      order:          c.order,
    }));

    const budget  = d.budget  || 0;
    const revenue = d.revenue || 0;

    return {
      tmdb_id:               d.id,
      title:                 d.title,
      overview:              d.overview ?? null,
      release_date:          d.release_date ?? null,
      budget,
      revenue,
      runtime:               d.runtime ?? null,
      popularity:            d.popularity ?? 0,
      vote_average:          d.vote_average ?? 0,
      vote_count:            d.vote_count ?? 0,
      poster_path:           d.poster_path ?? null,
      original_language:     d.original_language ?? null,
      belongs_to_collection: d.belongs_to_collection ?? null,

      genres:     (d.genres ?? []).map((g) => g.name),
      genre_ids:  (d.genres ?? []).map((g) => g.id),

      production_companies: (d.production_companies ?? []).map((c) => ({
        name: c.name,
        id:   c.id,
      })),
      production_countries: (d.production_countries ?? []).map((c) => ({
        iso_3166_1: c.iso_3166_1,
        name:       c.name,
      })),

      director: director
        ? { name: director.name, tmdb_person_id: director.id }
        : null,
      cast_top5,

      mpaa_rating:  extractMpaaRating(releaseDates),
      trailer_url:  extractTrailerUrl(videos),

      market:        detectMarket(d.production_countries, d.original_language),
      status:        mapStatus(d.status),
      budget_inferred: false,
      budget_source:   budget > 0 ? 'tmdb' : null,
    };
  } catch (err) {
    console.error('[TMDB ERROR] getFilmDetails:', err.message);
    return null;
  }
}

/* ------------------------------------------------------------------ */
/* 3. getPersonFilmography(tmdb_person_id)                            */
/* ------------------------------------------------------------------ */

export async function getPersonFilmography(tmdbPersonId) {
  try {
    const data = await tmdbGet(`/person/${tmdbPersonId}/movie_credits`);

    const as_cast = (data.cast ?? [])
      .filter((c) => c.order <= 2)
      .map((c) => ({
        tmdb_id:      c.id,
        title:        c.title,
        release_date: c.release_date ?? null,
        revenue:      c.revenue ?? 0,
        budget:       c.budget ?? 0,
        character:    c.character ?? null,
        order:        c.order,
      }));

    const as_crew = (data.crew ?? [])
      .filter((c) => c.job === 'Director')
      .map((c) => ({
        tmdb_id:      c.id,
        title:        c.title,
        release_date: c.release_date ?? null,
        revenue:      c.revenue ?? 0,
        budget:       c.budget ?? 0,
      }));

    return { as_cast, as_crew };
  } catch (err) {
    console.error('[TMDB ERROR] getPersonFilmography:', err.message);
    return null;
  }
}

/* ------------------------------------------------------------------ */
/* 4. getUpcomingFilms(market)                                        */
/* ------------------------------------------------------------------ */

async function fetchDiscoverPage(params, page) {
  const data = await tmdbGet('/discover/movie', { ...params, page });
  return data.results ?? [];
}

export async function getUpcomingFilms(market) {
  try {
    const baseParams = {
      'primary_release_date.gte': today(),
      'primary_release_date.lte': dateOffset(18),
      sort_by:                    'popularity.desc',
      include_adult:              false,
      language:                   'en-US',
      ...marketDiscoverParams(market),
    };

    let results;
    if (market === 'european') {
      // European: fetch non-English European language films + UK English separately.
      const [p1, p2, gbP1] = await Promise.all([
        fetchDiscoverPage(baseParams, 1),
        fetchDiscoverPage(baseParams, 2),
        fetchDiscoverPage(
          { ...baseParams, with_original_language: undefined, with_origin_country: 'GB' },
          1
        ),
      ]);
      const seen = new Set();
      results = [];
      for (const r of [...p1, ...p2, ...gbP1]) {
        if (!seen.has(r.id)) { seen.add(r.id); results.push(r); }
      }
    } else {
      const [p1, p2] = await Promise.all([
        fetchDiscoverPage(baseParams, 1),
        fetchDiscoverPage(baseParams, 2),
      ]);
      results = [...p1, ...p2];
    }

    return results.slice(0, 50).map((r) => ({
      tmdb_id:      r.id,
      title:        r.title,
      release_date: r.release_date ?? null,
      poster_path:  r.poster_path ?? null,
      popularity:   r.popularity ?? 0,
      overview:     r.overview ?? null,
    }));
  } catch (err) {
    console.error('[TMDB ERROR] getUpcomingFilms:', err.message);
    return null;
  }
}

/* ------------------------------------------------------------------ */
/* 5. getReleasedFilmsForComps(genre_ids, market, budget)             */
/* ------------------------------------------------------------------ */

// budget is accepted for future budget-proximity filtering but not yet used in the TMDB query.
export async function getReleasedFilmsForComps(genre_ids, market, _budget) {
  try {
    const primaryGenre = Array.isArray(genre_ids) ? genre_ids[0] : genre_ids;
    const params = {
      with_genres:                primaryGenre,
      'primary_release_date.lte': today(),
      'primary_release_date.gte': dateOffset(0, -8),
      sort_by:                    'revenue.desc',
      'vote_count.gte':           100,
      include_adult:              false,
      language:                   'en-US',
      ...marketDiscoverParams(market),
    };

    const [p1, p2] = await Promise.all([
      fetchDiscoverPage(params, 1),
      fetchDiscoverPage(params, 2),
    ]);

    return [...p1, ...p2].slice(0, 30).map((r) => ({
      tmdb_id:      r.id,
      title:        r.title,
      release_date: r.release_date ?? null,
      revenue:      r.revenue ?? 0,
      budget:       r.budget ?? 0,
      genre_ids:    r.genre_ids ?? [],
      poster_path:  r.poster_path ?? null,
    }));
  } catch (err) {
    console.error('[TMDB ERROR] getReleasedFilmsForComps:', err.message);
    return null;
  }
}

/* ------------------------------------------------------------------ */
/* Health check                                                       */
/* ------------------------------------------------------------------ */

// Country code per market for watch-provider lookups
const MARKET_COUNTRY = {
  hollywood: 'US', bollywood: 'IN', korean: 'KR', japanese: 'JP', european: 'GB',
};

/**
 * Fetch streaming / rental / purchase availability from TMDB.
 * Returns { flatrate, rent, buy, link, country } or null on failure.
 */
export async function getWatchProviders(tmdbId, market = 'hollywood') {
  try {
    const data    = await tmdbGet(`/movie/${tmdbId}/watch/providers`);
    const country = MARKET_COUNTRY[market] || 'US';
    const region  = (data.results || {})[country] || {};
    return {
      country,
      flatrate: (region.flatrate || []).map(p => ({
        provider_id:   p.provider_id,
        provider_name: p.provider_name,
        logo_path:     p.logo_path,
        logo_url:      p.logo_path ? `https://image.tmdb.org/t/p/original${p.logo_path}` : null,
      })),
      rent: (region.rent || []).map(p => ({
        provider_id:   p.provider_id,
        provider_name: p.provider_name,
        logo_path:     p.logo_path,
        logo_url:      p.logo_path ? `https://image.tmdb.org/t/p/original${p.logo_path}` : null,
      })),
      buy: (region.buy || []).map(p => ({
        provider_id:   p.provider_id,
        provider_name: p.provider_name,
        logo_path:     p.logo_path,
        logo_url:      p.logo_path ? `https://image.tmdb.org/t/p/original${p.logo_path}` : null,
      })),
      link: region.link || null,
    };
  } catch {
    return null;
  }
}

export async function testConnection() {
  try {
    await tmdbGet('/configuration');
    return { ok: true };
  } catch (err) {
    return { ok: false, error: err.message };
  }
}

File: backend/services/tmdb.py
import os
from datetime import datetime, timedelta, timezone
import httpx
from dotenv import load_dotenv

load_dotenv()

BASE = 'https://api.themoviedb.org/3'

BOLLYWOOD_LANGS = {'hi', 'te', 'ta', 'ml', 'kn'}
EUROPEAN_COUNTRIES = {'GB', 'FR', 'ES', 'IT', 'DE', 'SE', 'DK', 'NO', 'NL'}

MARKET_COUNTRY = {
    'hollywood': 'US', 'bollywood': 'IN', 'korean': 'KR',
    'japanese': 'JP', 'european': 'GB',
}


def _build_request():
    bearer = os.getenv('TMDB_READ_ACCESS_TOKEN')
    api_key = os.getenv('TMDB_API_KEY')
    if bearer:
        return {'Authorization': f'Bearer {bearer}', 'Content-Type': 'application/json'}, {}
    if api_key:
        return {}, {'api_key': api_key}
    raise ValueError('No TMDB credentials: set TMDB_READ_ACCESS_TOKEN or TMDB_API_KEY in .env')


async def _tmdb_get(pathname, params=None):
    headers, extra = _build_request()
    merged = {**(extra), **(params or {})}
    merged = {k: v for k, v in merged.items() if v is not None and v != ''}
    url = f'{BASE}{pathname}'
    async with httpx.AsyncClient(timeout=15.0) as client:
        res = await client.get(url, headers=headers, params=merged)
        res.raise_for_status()
        return res.json()


def _today():
    return datetime.now(timezone.utc).strftime('%Y-%m-%d')


def _date_offset(months=0, years=0):
    d = datetime.now(timezone.utc)
    m = d.month + months
    y = d.year + years + (m - 1) // 12
    m = ((m - 1) % 12) + 1
    return d.replace(year=y, month=m).strftime('%Y-%m-%d')


def _detect_market(production_countries, original_language):
    countries = [c.get('iso_3166_1', '') for c in (production_countries or [])]
    lang = original_language or ''
    if 'US' in countries and lang == 'en':
        return 'hollywood'
    if 'IN' in countries and lang in BOLLYWOOD_LANGS:
        return 'bollywood'
    if 'KR' in countries:
        return 'korean'
    if 'JP' in countries:
        return 'japanese'
    if any(c in EUROPEAN_COUNTRIES for c in countries):
        return 'european'
    if 'US' in countries:
        if lang in BOLLYWOOD_LANGS:
            return 'bollywood'
        if lang == 'ko':
            return 'korean'
        if lang == 'ja':
            return 'japanese'
        if lang in ('fr', 'de', 'es', 'it'):
            return 'european'
    return 'hollywood'


def _map_status(tmdb_status):
    s = (tmdb_status or '').lower()
    if s == 'released':
        return 'released'
    if s in ('in production', 'post production'):
        return 'in_production'
    return 'upcoming'


def _extract_mpaa(release_dates):
    results = (release_dates or {}).get('results', [])
    us = next((r for r in results if r.get('iso_3166_1') == 'US'), None)
    if us:
        cert = next(
            (rd.get('certification') for rd in us.get('release_dates', [])
             if rd.get('type') in (2, 3) and rd.get('certification')),
            None
        )
        if cert:
            return cert
        cert = next(
            (rd.get('certification') for rd in us.get('release_dates', [])
             if rd.get('certification')),
            None
        )
        if cert:
            return cert
    for entry in results:
        cert = next(
            (rd.get('certification') for rd in entry.get('release_dates', [])
             if rd.get('certification')),
            None
        )
        if cert:
            return cert
    return None


def _extract_trailer(videos):
    results = (videos or {}).get('results', [])
    official = next(
        (v for v in results if v.get('type') == 'Trailer' and v.get('official') and v.get('site') == 'YouTube'),
        None
    )
    if official:
        return f"https://www.youtube.com/watch?v={official['key']}"
    any_yt = next(
        (v for v in results if v.get('type') == 'Trailer' and v.get('site') == 'YouTube'),
        None
    )
    if any_yt:
        return f"https://www.youtube.com/watch?v={any_yt['key']}"
    return None


def _market_discover_params(market):
    mapping = {
        'bollywood': {'with_original_language': 'hi|te|ta|ml|kn'},
        'korean':    {'with_original_language': 'ko'},
        'japanese':  {'with_original_language': 'ja'},
        'european':  {'with_original_language': 'fr|de|es|it'},
        'hollywood': {'with_original_language': 'en'},
    }
    return mapping.get(market, {'with_original_language': 'en'})


async def _fetch_discover_page(params, page):
    data = await _tmdb_get('/discover/movie', {**params, 'page': page})
    return data.get('results', [])


async def search_film(query):
    try:
        data = await _tmdb_get('/search/movie', {'query': query, 'include_adult': False})
        return [
            {'tmdb_id': r['id'], 'title': r['title'],
             'release_date': r.get('release_date'), 'poster_path': r.get('poster_path'),
             'overview': r.get('overview')}
            for r in (data.get('results') or [])[:5]
        ]
    except Exception as e:
        print(f'[TMDB ERROR] search_film: {e}')
        return None


async def get_film_details(tmdb_id):
    try:
        d = await _tmdb_get(f'/movie/{tmdb_id}',
                            {'append_to_response': 'credits,release_dates,videos'})
        credits = d.get('credits', {})
        release_dates = d.get('release_dates', {})
        videos = d.get('videos', {})

        director_raw = next((c for c in credits.get('crew', []) if c.get('job') == 'Director'), None)
        cast_top5 = [
            {'name': c['name'], 'tmdb_person_id': c['id'], 'order': c['order']}
            for c in credits.get('cast', [])[:5]
        ]
        prod_countries = d.get('production_countries', [])
        orig_lang = d.get('original_language')

        return {
            'tmdb_id':               d['id'],
            'title':                 d['title'],
            'overview':              d.get('overview'),
            'release_date':          d.get('release_date'),
            'budget':                d.get('budget') or 0,
            'revenue':               d.get('revenue') or 0,
            'runtime':               d.get('runtime'),
            'popularity':            d.get('popularity', 0),
            'vote_average':          d.get('vote_average', 0),
            'vote_count':            d.get('vote_count', 0),
            'poster_path':           d.get('poster_path'),
            'original_language':     orig_lang,
            'belongs_to_collection': d.get('belongs_to_collection'),
            'genres':                [g['name'] for g in d.get('genres', [])],
            'genre_ids':             [g['id']   for g in d.get('genres', [])],
            'production_companies':  [{'name': c['name'], 'id': c['id']}
                                      for c in d.get('production_companies', [])],
            'production_countries':  [{'iso_3166_1': c['iso_3166_1'], 'name': c['name']}
                                      for c in prod_countries],
            'director':              {'name': director_raw['name'],
                                      'tmdb_person_id': director_raw['id']} if director_raw else None,
            'cast_top5':             cast_top5,
            'mpaa_rating':           _extract_mpaa(release_dates),
            'trailer_url':           _extract_trailer(videos),
            'market':                _detect_market(prod_countries, orig_lang),
            'status':                _map_status(d.get('status')),
            'budget_inferred':       False,
            'budget_source':         'tmdb' if (d.get('budget') or 0) > 0 else None,
        }
    except Exception as e:
        print(f'[TMDB ERROR] get_film_details: {e}')
        return None


async def get_person_filmography(tmdb_person_id):
    try:
        data = await _tmdb_get(f'/person/{tmdb_person_id}/movie_credits')
        as_cast = [
            {'tmdb_id': c['id'], 'title': c['title'], 'release_date': c.get('release_date'),
             'revenue': c.get('revenue', 0), 'budget': c.get('budget', 0),
             'character': c.get('character'), 'order': c['order']}
            for c in data.get('cast', []) if c.get('order', 99) <= 2
        ]
        as_crew = [
            {'tmdb_id': c['id'], 'title': c['title'], 'release_date': c.get('release_date'),
             'revenue': c.get('revenue', 0), 'budget': c.get('budget', 0)}
            for c in data.get('crew', []) if c.get('job') == 'Director'
        ]
        return {'as_cast': as_cast, 'as_crew': as_crew}
    except Exception as e:
        print(f'[TMDB ERROR] get_person_filmography: {e}')
        return None


async def get_upcoming_films(market):
    try:
        base = {
            'primary_release_date.gte': _today(),
            'primary_release_date.lte': _date_offset(months=18),
            'sort_by': 'popularity.desc',
            'include_adult': False,
            'language': 'en-US',
            **_market_discover_params(market),
        }
        import asyncio
        if market == 'european':
            gb_params = {k: v for k, v in base.items() if k != 'with_original_language'}
            gb_params['with_origin_country'] = 'GB'
            p1, p2, gb = await asyncio.gather(
                _fetch_discover_page(base, 1),
                _fetch_discover_page(base, 2),
                _fetch_discover_page(gb_params, 1),
            )
            seen, results = set(), []
            for r in [*p1, *p2, *gb]:
                if r['id'] not in seen:
                    seen.add(r['id'])
                    results.append(r)
        else:
            p1, p2 = await asyncio.gather(
                _fetch_discover_page(base, 1),
                _fetch_discover_page(base, 2),
            )
            results = [*p1, *p2]

        return [
            {'tmdb_id': r['id'], 'title': r['title'], 'release_date': r.get('release_date'),
             'poster_path': r.get('poster_path'), 'popularity': r.get('popularity', 0),
             'overview': r.get('overview')}
            for r in results[:50]
        ]
    except Exception as e:
        print(f'[TMDB ERROR] get_upcoming_films: {e}')
        return None


async def get_released_films_for_comps(genre_ids, market, _budget=None):
    try:
        primary = genre_ids[0] if isinstance(genre_ids, list) else genre_ids
        params = {
            'with_genres': primary,
            'primary_release_date.lte': _today(),
            'primary_release_date.gte': _date_offset(years=-8),
            'sort_by': 'revenue.desc',
            'vote_count.gte': 100,
            'include_adult': False,
            'language': 'en-US',
            **_market_discover_params(market),
        }
        import asyncio
        p1, p2 = await asyncio.gather(
            _fetch_discover_page(params, 1),
            _fetch_discover_page(params, 2),
        )
        return [
            {'tmdb_id': r['id'], 'title': r['title'], 'release_date': r.get('release_date'),
             'revenue': r.get('revenue', 0), 'budget': r.get('budget', 0),
             'genre_ids': r.get('genre_ids', []), 'poster_path': r.get('poster_path')}
            for r in ([*p1, *p2])[:30]
        ]
    except Exception as e:
        print(f'[TMDB ERROR] get_released_films_for_comps: {e}')
        return None


async def get_watch_providers(tmdb_id, market='hollywood'):
    try:
        data = await _tmdb_get(f'/movie/{tmdb_id}/watch/providers')
        country = MARKET_COUNTRY.get(market, 'US')
        region = (data.get('results') or {}).get(country, {})

        def _map_providers(lst):
            return [
                {'provider_id': p['provider_id'], 'provider_name': p['provider_name'],
                 'logo_path': p.get('logo_path'),
                 'logo_url': f"https://image.tmdb.org/t/p/original{p['logo_path']}"
                             if p.get('logo_path') else None}
                for p in (lst or [])
            ]

        return {
            'country':  country,
            'flatrate': _map_providers(region.get('flatrate')),
            'rent':     _map_providers(region.get('rent')),
            'buy':      _map_providers(region.get('buy')),
            'link':     region.get('link'),
        }
    except Exception:
        return None


async def test_connection():
    try:
        await _tmdb_get('/configuration')
        return {'ok': True}
    except Exception as e:
        return {'ok': False, 'error': str(e)}
